event/src/unvalidated/builder.rs

```diff
@@ -0,0 +1,277 @@
+    }
+
+    /// Create builder for init event with data
+    pub fn with_data<T>(self, data: T) -> InitWithDataBuilder<T> {
+        InitWithDataBuilder {
+            data,
+            controllers: self.controllers,
+            sep: self.sep,
+            additional: self.additional,
+        }
+    }
+}
+
+/// Builder for init event with data
+pub struct InitWithDataBuilder<D> {
+    data: D,
+    controllers: Vec<String>,
+    sep: String,
+    additional: HashMap<String, Value>,
+}
+
+impl<D> Additional for InitWithDataBuilder<D> {
+    fn with_additional(mut self, key: String, value: Value) -> Self {
+        self.additional.insert(key, value);
+        self
+    }
+}
+
+impl<D> Controllers for InitWithDataBuilder<D> {
+    fn with_controller(mut self, controller: String) -> Self {
+        self.controllers.push(controller);
+        self
+    }
+}
+
+impl<D> Sep for InitWithDataBuilder<D> {
+    fn with_sep(mut self, sep: String) -> Self {
+        self.sep = sep;
+        self
+    }
+}
+
+impl<D> InitWithDataBuilder<D> {
+    /// Build unsigned init event with data
+    pub async fn build(self) -> anyhow::Result<InitPayload<D>> {
+        if self.controllers.is_empty() {
+            anyhow::bail!("controllers must not be empty");
+        }
+        if self.sep.is_empty() {
+            anyhow::bail!("sep must not be empty");
+        }
+        let header = InitHeader::new(self.controllers, self.sep, self.additional);
+        let payload = InitPayload::new(header, Some(self.data));
+        Ok(payload)
+    }
+}
+
+/// Builder for data event
+pub struct DataBuilder<T> {
+    id: Cid,
+    prev: Cid,
+    data: T,
+    controllers: Vec<String>,
+    sep: String,
+    additional: HashMap<String, Value>,
+}
+
+impl<T> Additional for DataBuilder<T> {
+    fn with_additional(mut self, key: String, value: Value) -> Self {
+        self.additional.insert(key, value);
+        self
+    }
+}
+
+impl<T> Controllers for DataBuilder<T> {
+    fn with_controller(mut self, controller: String) -> Self {
+        self.controllers.push(controller);
+        self
+    }
+}
+
+impl<T> Sep for DataBuilder<T> {
+    fn with_sep(mut self, sep: String) -> Self {
+        self.sep = sep;
+        self
+    }
+}
+
+impl<T> DataBuilder<T> {
+    /// Build data event
+    pub async fn build(self) -> anyhow::Result<DataPayload<T>> {
+        let header = DataHeader::new(self.controllers, self.additional);
+        let payload = DataPayload::new(self.id, self.prev, Some(header), self.data);
+        Ok(payload)
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::unvalidated::payload::Payload;
+
+    const TEST_CID: &str = "bafyreiaroclcgqih242byss6pneufencrulmeex2ttfdzefst67agwq3im";
+
+    #[tokio::test]
+    async fn should_build_init_event_without_data() {
+        let event = Builder::default()
+            .with_sep("sep".to_string())
+            .with_controller("controller".to_string())
+            .init()
+            .build()
```

Comment:
I would like the builder API to make it impossible to build events that are nonsensical. For example building a data event with a controller. I think this test is a good example of how its possible to create nonsensical events with the current API.

Since `init` can be called after `with_controller` this implies that `with_controller` can be called on a data event(In fact the DataBuilder struct implements the Controller trait and I am not sure why). I would like an builder API that makes you choose as the first thing the kind of event you plan to build. For example `Builder::init().with_controller()`  vs `Builder::data().with_data()` or similar. Thoughts?